enhancements/cert-manager/istio-csr-controller.md

```diff
@@ -0,0 +1,878 @@
+type TLSConfig struct {
+	// TrustDomain is the cluster's trust domain.
+	// +kubebuilder:default:="cluster.local"
+	// +optional
+	TrustDomain string `json:"trustDomain,omitempty"`
+
+	// RootCAFile is for setting the file location containing the root CA which is
+	// present in the configured IssuerRef. File should be made available using the
+	// Volume and VolumeMount options.
+	// +optional
+	RootCAFile string `json:"rootCAFile,omitempty"`
+
+	// CertificateDNSNames contains DNS names to be added to the certificate SAN.
+	// +kubebuilder:default:=["cert-manager-istio-csr.istio-system.svc"]
+	// +optional
+	CertificateDNSNames []string `json:"certificateDNSNames,omitempty"`
+
+	// IstiodCertificateDuration is the istio's certificate validity period.
+	// +kubebuilder:default:="1h"
+	// +optional
+	IstiodCertificateDuration time.Duration `json:"istiodCertificateDuration,omitempty"`
+
+	// IstiodCertificateRenewBefore is the ahead time to renew the istio's certificate before
+	// expiry.
+	// +kubebuilder:default:="30m"
+	// +optional
+	IstiodCertificateRenewBefore time.Duration `json:"istiodCertificateRenewBefore,omitempty"`
+
+	// IstiodPrivateKeySize is the key size to be for RSAKey.
+	// +kubebuilder:default:=2048
+	// +optional
+	IstiodPrivateKeySize int32 `json:"istiodPrivateKeySize,omitempty"`
+}
+
+// ServerConfig is for configuring the server endpoint used by istio
+// for ontaining the certificates.
+type ServerConfig struct {
+	// Address to serve istio-csr gRPC service.
+	// +kubebuilder:default:="0.0.0.0"
+	// +optional
+	Address string `json:"address,omitempty"`
+
+	// Port to serve istio-csr gRPC service.
+	// +kubebuilder:default:="6443"
+	// +optional
+	Port string `json:"port,omitempty"`
+
+	// CertificateKeySize is the server's serving certificate's key size.
+	// +kubebuilder:default:=2048
+	// +optional
+	CertificateKeySize int32 `json:"certificateKeySize,omitempty"`
+
+	// SignatureAlgorithm is the signature algorithm to use when generating
+	// private keys. At present only RSA and ECDSA are supported.
+	// +kubebuilder:default:="RSA"
+	// +kubebuilder:validation:Enum:="RSA";"ECDSA"
+	// +optional
+	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`
+
+	// MaxCertificateDuration is the maximum validity duration that can be
+	// requested for a certificate.
+	// +kubebuilder:default:="1h"
+	// +optional
+	MaxCertificateDuration time.Duration `json:"certificateDuration,omitempty"`
+}
+
+// IstioConfig is for configuring the istio specifics.
+type IstioConfig struct {
+	// Revisions are the istio revisions that are currently installed in the cluster.
+	// Changing this field will modify the DNS names that will be requested for
+	// the istiod certificate.
+	// +kubebuilder:default:=["default"]
+	// +optional
+	Revisions []string `json:"revisions,omitempty"`
+}
+
+// IstioCSRStatus is the most recently observed status of the IstioCSR.
+type IstioCSRStatus struct {
```

Comment:
A re-look at status fields of operator and controller will be needed as per new context.

Reply:
@TrilokGeer I think offline discussion we had to aggregate the state of all operands managed by the operator will be a different proposal, and here the `IstioCSRStatus` will just hold the details of deployed `istio-csr` operand. Please let me know WDYT?